### PR TITLE
Rewrite tests with jest

### PR DESCRIPTION
**Pull request by kachkaev**
Hey @jfairbank! Following your [approval](https://github.com/jfairbank/run-elm/pull/5#issuecomment-360482736) to use a testing framework, I did a bit more refactoring and was able to switch to jest.

Test case projects are now located inside `test/integration/case-projects`, which gives us some room to write unit tests inside `test/unit` in future. Any file called `test/**.test.js` is automatically picked by jest and is seamlessly babelified before execution.

As a by-product of switching to jest I also moved `index.js` to `src` and put `*.elm.template` files to the same directory. Because of this change, references to `package.json` and elm templates a valid both before and after `yarn build`. Moreover, `lib` can be now completely ignored by git, because elm templates are copied to that folder on build. All this does not produce any immediate benefits, but I guess they'll come after you split `index.js` into smaller modules and decide to test them separately.

Unfortunately, the tests are still running concurrently, but this might change if I figure out how to overcome https://github.com/facebook/jest/issues/5425. Not a big worry anyway.

One more caveat is that coverage reports from `yarn test --coverage` are currently empty, because instead of testing raw modules inside `src`, we execute the pre-built code from `lib`. To be honest, I'm not sure how to get coverage reports in this scenario, but if you have ideas or know an example, I'll be happy to work on that!

Curios to know what you think 💭

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,10 +13,10 @@
     "run-elm": "lib/index.js"
   },
   "scripts": {
-    "lint": "eslint index.js",
+    "lint": "eslint src test --ignore-pattern elm-stuff",
     "pretest": "npm run build",
-    "test": "node tests/run-tests.js",
-    "build": "babel index.js --out-dir lib",
+    "test": "jest \\.test.js",
```
Comment (jfairbank) [inline package.json]: I don't think we need the `\\.test.js` part here.

File: test/integration/index.test.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      try {
+        result = await execFile(runElmPath, input, { cwd: projectDir });
+      } catch (e) {
+        const { code, stderr } = e;
+        done.fail(`process timeout or non-zero exit code ${code}${stderr ? `: ${stderr}` : ''}`);
+        return;
+      }
+      expect(result.stdout).toEqual(expectedOutput);
+      expect(result.stderr).toEqual('');
+      done();
```
Comment (jfairbank) [inline test/integration/index.test.js]: Do we need the `done` callback if we're already using promises/async/await?

File: test/integration/index.test.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      const inputPath = resolve(projectDir, 'input.txt');
+      const input = (await readFile(inputPath, 'utf-8')).trim().split(' ');
+      const expectedOutput = await readFile(resolve(projectDir, 'output.txt'), 'utf-8');
+
+      let result;
+      try {
+        result = await execFile(runElmPath, input, { cwd: projectDir });
+      } catch (e) {
+        const { code, stderr } = e;
+        done.fail(`process timeout or non-zero exit code ${code}${stderr ? `: ${stderr}` : ''}`);
```
Comment (jfairbank) [inline test/integration/index.test.js]: Would throwing an error here with this message be the same so we can avoid pulling in the `done` callback?

**Review by jfairbank — CHANGES_REQUESTED** [review]
Looks great @kachkaev! Thanks for putting this together.

I had a couple questions and a minor nit.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,10 @@
+## next
+
+### Other Changes
+
+* switch to jest testing framework ([#7](https://github.com/jfairbank/run-elm/pull/7))
```
Comment (jfairbank) [inline CHANGELOG.md]: I don't think it's important to include this in the CHANGELOG since it's internal.

**Review by jfairbank — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (kachkaev) [reply package.json]: You're right – [default value of `testMatch`](https://facebook.github.io/jest/docs/en/configuration.html#testmatch-array-string) suits us.

File: test/integration/index.test.js (hunk 3 above)
Comment (kachkaev) [reply test/integration/index.test.js]: Actually yes, that's doable. Thanks for spotting!

File: test/integration/index.test.js (hunk 2 above)
Comment (kachkaev) [reply test/integration/index.test.js]: True, we don't. Nevertheless, keeping `try/catch` and `expect` outside `try/catch` seems still reasonable. In theory, we can simplify the test down to just six lines:

```js
    test(`correctly works for case project \`${projectName}\``, async () => {
      const inputPath = resolve(projectDir, 'input.txt');
      const input = (await readFile(inputPath, 'utf-8')).trim().split(' ');
      const expectedOutput = await readFile(resolve(projectDir, 'output.txt'), 'utf-8');

      const result = await execFile(runElmPath, input, { cwd: projectDir });
      expect(result.stdout).toEqual(expectedOutput);
      expect(result.stderr).toEqual('');
    }, 30000);
``` 

However, in this case an error triggered by `exeFile` will be quite cryptic:

```
ChildProcessError: spawn /path/to/run-elm/lib/index.js ENOENT `/path/to/run-elm/lib/index.js Main.elm hello world` (exited with error code ENOENT)
```

Extra `let result;`, `try/catch` and `throw` look less clean in the code, but they provide better feedback on test failures.

File: CHANGELOG.md (hunk 4 above)
Comment (kachkaev) [reply CHANGELOG.md]: OK, I won't argue :–) Removing that 👍

**Comment by kachkaev** [comment]
Done! I also tried `test.concurrent` after removing `done`, but that did not work well. Because too many `elm-make` instances are spawned simultaneously,  each test runs longer and 30-second timeouts are sometimes triggered when `elm-stuff` is empty.

**Review by jfairbank — APPROVED** [review]

**Comment by jfairbank** [comment]
Awesome! Thanks, @kachkaev!
